siac/parse.go

```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+var ErrUnableToParseSize = errors.New("unable to parse size")
+
+// parseSize converts strings of form 10GB to a size in bytes, if it fails an
+// error is returned.
+func parseSize(strSize string) (string, error) {
+	// If the value is at least 2 or 3 characters it may have a suffix.
+	// To verify it could be a suffix one of the characters are checked to see
+	// it is outside the ascii range for numbers (48-57),
+	// letters are (65-90, 97-122).
+	var suffix string
+	var size int
+	if len(strSize) > 3 && strSize[len(strSize)-3] > 57 {
+		suffix = strSize[len(strSize)-3:] // KiB, MiB, GiB, TiB
+		_, err := fmt.Sscan(strSize[:len(strSize)-3], &size)
+		if err != nil {
+			return "", ErrUnableToParseSize
+		}
+	} else if len(strSize) > 2 && strSize[len(strSize)-2] > 57 {
+		suffix = strSize[len(strSize)-2:] // KB, MB, GB, TB
+		_, err := fmt.Sscan(strSize[:len(strSize)-2], &size)
+		if err != nil {
+			return "", ErrUnableToParseSize
+		}
+	} else {
+		ret := fmt.Sprintf("%d", strSize)
+		if ret == "" {
+			return "", ErrUnableToParseSize
+		}
+		return ret, nil
+	}
+
+	switch strings.ToLower(suffix) {
+	case "tb":
+		size *= 1000
+		fallthrough
+	case "gb":
+		size *= 1000
+		fallthrough
+	case "mb":
+		size *= 1000
+		fallthrough
+	case "kb":
+		size *= 1000
+		break
```

Comment:
this is a clever use of `fallthrough`, but the conversion can be done more compactly.